modules/stochastic_tools/src/reporters/ActiveLearningGPDecision.C

```diff
@@ -0,0 +1,240 @@
+{
+  _gp_sto[0] = getSurrogateModel<GaussianProcess>("gp_evaluator").evaluate(row, _gp_sto[1]);
+  _gp_sto[1] = retrain ? 0.0 : _gp_sto[1];
+  bool lf_indicator = learningFunction(_gp_sto[0],
+                                       _gp_sto[1],
+                                       _learning_function,
+                                       *_learning_function_parameter,
+                                       _learning_function_threshold);
+  if (_flag_sample[local_ind] == true)
+    _flag_sample[local_ind] = false;
+  if (lf_indicator)
+  {
+    val = _gp_sto[0];
+    _decision[local_ind] = false;
+  }
+  else
+  {
+    ++_track_gp_fails;
+    _flag_sample[local_ind] = true;
+  }
+  _gp_mean_comm[local_ind] = _gp_sto[0];
+  _gp_std_comm[local_ind] = _flag_sample[0] == true ? 0.0 : _gp_sto[1];
+  return val;
+}
+
+void
+ActiveLearningGPDecision::transmitOutput(const DenseMatrix<Real> & inputs_comm,
+                                         const std::vector<Real> & gp_mean_comm,
+                                         const std::vector<Real> & gp_std_comm)
+{
+  for (dof_id_type ss = 0; ss < _sampler.getNumberOfRows(); ++ss)
+  {
+    for (unsigned int j = 0; j < _sampler.getNumberOfCols(); ++j)
+      _inputs[ss][j] = inputs_comm(ss, j);
+    _gp_mean[ss] = gp_mean_comm[ss];
+    _gp_std[ss] = gp_std_comm[ss];
+  }
+}
+
+bool
+ActiveLearningGPDecision::needSample(const std::vector<Real> & row,
+                                     dof_id_type local_ind,
+                                     dof_id_type,
+                                     Real & val)
+{
+  _gp_sto.resize(2);
+  _output_comm.resize(1);
+  _gp_mean_comm.resize(1);
+  _gp_std_comm.resize(1);
+  DenseMatrix<Real> inputs_comm(_sampler.getNumberOfRows(), _sampler.getNumberOfCols());
+  // Storing all the inputs and outputs (required for each time step of the mainApp)
+  for (dof_id_type ss = _sampler.getLocalRowBegin(); ss < _sampler.getLocalRowEnd(); ++ss)
+  {
+    const auto data = _sampler.getNextLocalRow();
+    for (unsigned int j = 0; j < _sampler.getNumberOfCols(); ++j)
+      inputs_comm(ss, j) = data[j];
+  }
+  _output_comm[local_ind] = val;
+  _local_comm.sum(inputs_comm.get_values());
+  _local_comm.allgather(_output_comm);
+  if (_step <= _n_train) // Wait until all the training data is generated
+  {
+    if (_step > 2)
+      setupData(_output_comm, _inputs_prev);
+    if (_step == _n_train) // Once training data is generated, train the GP
+    {
+      if (local_ind == 0)
+        _al_gp->reTrain(_inputs_sto, _outputs_sto);
+      // Setting up variables and making decisions
+      val = facilitateDecision(row, local_ind, val, true);
+      _local_comm.allgather(_gp_mean_comm);
+      _local_comm.allgather(_gp_std_comm);
+      transmitOutput(inputs_comm, _gp_mean_comm, _gp_std_comm);
+    }
+    // Start tracking the GP failures until a user-specified batch size is met
+    if (_track_gp_fails >= _allowed_gp_fails)
```

Comment:
Do we care about GP failures in the initial training process? I thought this was only to give the GP a little initial training.

Reply:
Yes, because this step is when we start making GP predictions first before calling the subApp.